packages/data/src/notes/selectors.js

```diff
@@ -1,7 +1,18 @@
-export const getNotes = ( state, query ) => {
-	const noteIds = state.noteQueries[ JSON.stringify( query ) ] || [];
-	return noteIds.map( ( id ) => state.notes[ id ] );
-};
+/**
+ * External dependencies
+ */
+import createSelector from 'rememo';
+
+export const getNotes = createSelector(
+	( state, query ) => {
+		const noteIds = state.noteQueries[ JSON.stringify( query ) ] || [];
+		return noteIds.map( ( id ) => state.notes[ id ] );
+	},
+	( state, query ) => [
+		state.noteQueries[ JSON.stringify( query ) ],
+		state.notes,
+	]
+);
```

Comment:
Make use of the `createSelector` function, this memoizes the return value.
The selector only gets updated when the second parameter function is updated (line 12 & 13). 
In this case the selector is run when:
- the `noteQueries` array is updated for the specific `query`
- When the `notes` state is updated, needed for when a note is dismissed (deleted).

When the above two are not updated it will return the last cached result, which would contain the same object reference if that is what it is returning. (see the [wordpress/core-data](https://github.com/WordPress/gutenberg/blob/master/packages/core-data/src) library for more examples.

Reply:
I like this a lot. Redux is the perfect layer to do this kind of biz-logic caching and means we can just forget about it at the component layer. Good stuff!